waku/v2/node/waku_node.nim

```diff
@@ -292,63 +292,71 @@ proc connectToNodes*(node: WakuNode, nodes: seq[RemotePeerInfo] | seq[string], s
-    # TODO: improved error handling
+proc registerRelayDefaultHandler(node: WakuNode, topic: PubsubTopic) =
+  if node.wakuRelay.isSubscribed(topic):
     return
 
-  info "subscribe", topic=topic
-
-  proc defaultHandler(topic: string, data: seq[byte]) {.async, gcsafe.} =
-    # A default handler should be registered for all topics
-
-    let msg = WakuMessage.decode(data)
-    if msg.isErr():
-      # TODO: Add metric to track waku message decode errors
-      return
-
+  proc traceHandler(topic: PubsubTopic, data: seq[byte]) {.async, gcsafe.} =
     trace "waku.relay received",
       pubsubTopic=topic,
       hash=MultiHash.digest("sha2-256", data).expect("valid hash").data.buffer.to0xHex(), # TODO: this could be replaced by a message UID
       receivedTime=getNowInNanosecondTime()
 
-    # Notify mounted protocols of new message
-    if not node.wakuFilter.isNil():
-      await node.wakuFilter.handleMessage(topic, msg.value)
+    waku_node_messages.inc(labelValues = ["relay"])
+
+  proc filterHandler(topic: PubsubTopic, msg: WakuMessage) {.async, gcsafe.} =
+    if node.wakuFilter.isNil():
+      return
 
-    if not node.wakuArchive.isNil():
-      node.wakuArchive.handleMessage(topic, msg.value)
+    await node.wakuFilter.handleMessage(topic, msg)
 
-    waku_node_messages.inc(labelValues = ["relay"])
+  proc archiveHandler(topic: PubsubTopic, msg: WakuMessage) {.async, gcsafe.} =
+    if node.wakuArchive.isNil():
+      return
 
+    node.wakuArchive.handleMessage(topic, msg)
 
-  let wakuRelay = node.wakuRelay
 
-  if topic notin PubSub(wakuRelay).topics:
-    # Add default handler only for new topics
-    debug "Registering default handler", topic=topic
-    wakuRelay.subscribe(topic, defaultHandler)
+  let defaultHandler = proc(topic: PubsubTopic, data: seq[byte]) {.async, gcsafe.} =
+      let msg = WakuMessage.decode(data)
+      if msg.isErr():
+        return
+
+      await traceHandler(topic, data)
+      await filterHandler(topic, msg.value)
+      await archiveHandler(topic, msg.value)
+
+  node.wakuRelay.subscribe(topic, defaultHandler)
 
-  if handler.isSome():
-    debug "Registering handler", topic=topic
-    wakuRelay.subscribe(topic, handler.get())
 
-proc subscribe*(node: WakuNode, topic: PubsubTopic, handler: TopicHandler) =
+proc subscribe*(node: WakuNode, topic: PubsubTopic) =
+  if node.wakuRelay.isNil():
+    error "Invalid API call to `subscribe`. WakuRelay not mounted."
+    return
+
+  debug "subscribe", pubsubTopic= topic
+
+  node.registerRelayDefaultHandler(topic)
+
+proc subscribe*(node: WakuNode, topic: PubsubTopic, handler: WakuRelayHandler) =
   ## Subscribes to a PubSub topic. Triggers handler when receiving messages on
   ## this topic. TopicHandler is a method that takes a topic and some data.
-  ##
-  ## NOTE The data field SHOULD be decoded as a WakuMessage.
-  node.subscribe(topic, some(handler))
+  if node.wakuRelay.isNil():
+    error "Invalid API call to `subscribe`. WakuRelay not mounted."
```

Comment:
No, why:

> Applications should only perform panics. Per definition, a `panic` is a **non-maskable** exception. So, in general, it is a bad design to panic inside a library since users need help to handle the failure and act accordingly.
> 
> Additionally, protocol-related functions should not be part of this type's responsibility. The `WakuNode` type is a good example of one of the most problematic antipatterns: [the God class](https://linearb.io/blog/what-is-a-god-class/).

This is the next big and challenging refactoring in the code base, after the JSON-RPC/Rest APIs. There is a [huge technical debt pile](http://vincentdnl.com/static/76269aee1152cc12dd266ebd0ddf3480/ddf68/technical-debt.png) in the `waku_node` file.